sdks/java/io/kinesis/src/test/java/org/apache/beam/sdk/io/kinesis/KinesisIOIT.java

```diff
@@ -99,28 +116,101 @@ private void runRead() {
+    System.setProperty(SDKGlobalConfiguration.DISABLE_CERT_CHECKING_SYSTEM_PROPERTY, "true");
+    System.setProperty(SDKGlobalConfiguration.AWS_CBOR_DISABLE_SYSTEM_PROPERTY, "true");
+
+    // For some unclear reason localstack requires the timestamp in seconds
+    now = Instant.ofEpochMilli(Long.divideUnsigned(Instant.now().getMillis(), 1000));
+
+    localstackContainer =
+        new LocalStackContainer("0.11.3")
+            .withServices(LocalStackContainer.Service.KINESIS)
+            .withEnv("USE_SSL", "true")
+            .withStartupAttempts(3);
+    localstackContainer.start();
+
+    options.setAwsServiceEndpoint(
+        localstackContainer
+            .getEndpointConfiguration(LocalStackContainer.Service.KINESIS)
+            .getServiceEndpoint()
+            .replace("http", "https"));
+    options.setAwsKinesisRegion(
+        localstackContainer
+            .getEndpointConfiguration(LocalStackContainer.Service.KINESIS)
+            .getSigningRegion());
+    options.setAwsAccessKey(
+        localstackContainer.getDefaultCredentialsProvider().getCredentials().getAWSAccessKeyId());
+    options.setAwsSecretKey(
+        localstackContainer.getDefaultCredentialsProvider().getCredentials().getAWSSecretKey());
+    options.setNumberOfRecords(1000);
+    options.setNumberOfShards(1);
+    options.setAwsKinesisStream("beam_kinesis_test");
+    options.setAwsVerifyCertificate(false);
+    createStream(options.getAwsKinesisStream());
+  }
+
+  private static void createStream(String streamName) throws Exception {
+    AmazonKinesisClientBuilder clientBuilder = AmazonKinesisClientBuilder.standard();
+
+    clientBuilder.setCredentials(localstackContainer.getDefaultCredentialsProvider());
+    clientBuilder.setEndpointConfiguration(
+        localstackContainer.getEndpointConfiguration(LocalStackContainer.Service.KINESIS));
+
+    AmazonKinesis client = clientBuilder.build();
+
+    client.createStream(streamName, 1);
+    int repeats = 10;
+    for (int i = 0; i <= repeats; ++i) {
+      String streamStatus =
+          client.describeStream(streamName).getStreamDescription().getStreamStatus();
+      if ("ACTIVE".equals(streamStatus)) {
+        break;
+      }
+      if (i == repeats) {
+        throw new RuntimeException("Unable to initialize stream");
+      }
+      Thread.sleep(1000L);
+    }
+  }
+
+  /** Check whether pipeline options were provided. If not, use localstack container. */
+  private static boolean doUseLocalstack() {
+    return "aws-access-key".equals(options.getAwsAccessKey())
+        && "aws-secret-key".equals(options.getAwsSecretKey())
+        && "aws-kinesis-stream".equals(options.getAwsKinesisStream())
+        && "aws-kinesis-region".equals(options.getAwsKinesisRegion())
+        && options.getNumberOfShards() == 2
+        && options.getNumberOfRecords() == 1000
+        && options.getAwsServiceEndpoint() == null
+        && options.getAwsVerifyCertificate();
+  }
```

Comment:
Rather than repeating the default values here, you could create a fresh instance using `PipelineOptionsFactory.fromArgs().as(KinesisTestOptions.class)` and compare each field against it.

Reply:
Very good point, thanks!